Add `model_context` to `SelectorGroupChat` for enhanced speaker selection

Ethan0456 opened the pull request:
## Why are these changes needed?
This PR enhances the `SelectorGroupChat` class by introducing a new `model_context` parameter to support more context-aware speaker selection.

### Changes
- Added a `model_context: ChatCompletionContext | None` parameter to `SelectorGroupChat`.
- Defaulted to `UnboundedChatCompletionContext` when None is provided like `AssistantAgent`.
- Updated `_select_speaker` to prepend context messages from `model_context` to the main thread history.
- Refactored history construction into a helper method `construct_message_history`.

## Related issue number
Closes [Issue #6301](https://github.com/org/repo/issues/6301), enabling the group chat manager to utilize `model_context` for richer, more informed speaker selection decisions.

## Checks
- [x] I've included any doc changes needed for <https://microsoft.github.io/autogen/>. See <https://github.com/microsoft/autogen/blob/main/CONTRIBUTING.md> to build and test documentation locally.
- [x] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [x] I've made sure all auto checks have passed.

Ethan0456 commented:
Hi @ekzhu,

I’ve made some changes to use messages from `model_context` for speaker selection. For now, `BufferedChatCompletionContext` with a buffer size of 5 is set as the default for testing.  

Let me know if you have any feedback on the approach — and also curious which context class you'd prefer as the default.

ekzhu commented on python/packages/autogen-agentchat/src/autogen_agentchat/teams/_group_chat/_selector_group_chat.py:
```diff
@@ -79,13 +91,20 @@ def __init__(
+            self._model_context = model_context
+        else:
+            # TODO: finalize the best default context class
+            self._model_context = BufferedChatCompletionContext(buffer_size=5)
```
Let's use UnboundedChatCompletionContext

Ethan0456 replied:
Done! changed it to use `UnboundedChatCompletionContext`

ekzhu commented on python/packages/autogen-agentchat/src/autogen_agentchat/teams/_group_chat/_selector_group_chat.py:
```diff
@@ -101,15 +120,36 @@ async def save_state(self) -> Mapping[str, Any]:
         A key assumption is that the agent type is the same as the topic type, which we use as the agent name.
         """
+        # TODO: A hacky solution - Update model context from _message_thread at every speaker selection
+        # Add last BaseChatMessage to model context
```
Split the part of `handle_agent_response` in the base class that deal with updating message thread into a separate method: `update_message_thread`, which updates the message thread. 

https://github.com/microsoft/autogen/blob/d051da52c3a5f85889a6b89b30f0648dc00d1834/python/packages/autogen-agentchat/src/autogen_agentchat/teams/_group_chat/_base_group_chat_manager.py#L145-L153

Then, in `SelectorGroupChatManager`, override this method to also update the message context.

ekzhu replied:
Same with `handle_start`, we also need to update model context there. https://github.com/microsoft/autogen/blob/d051da52c3a5f85889a6b89b30f0648dc00d1834/python/packages/autogen-agentchat/src/autogen_agentchat/teams/_group_chat/_base_group_chat_manager.py#L118-L119

Ethan0456 replied:
Done!

Created a `update_message_thread` method in `BaseGroupChatManager` to update the message thread and overrode it in `SelectorGroupChatManager` to update the `model_context`

ekzhu commented:
@Ethan0456 I realized that https://github.com/microsoft/autogen/pull/6350/ may be doing a similar thing to this PR but from message thread point of view. Let's pause this PR for now and let's see if we can address context size problem using #6350 first.

SongChiYoung commented:
@ekzhu @Ethan0456 

As an AutoGen user who has been eagerly looking forward to this PR, I wanted to share my thoughts in detail. It's a bit long, but I hope it's clear. I would appreciate any feedback after reading.

#### **Community Need**  
Based on ongoing community feedback, I believe there is a clear need for internal message summarization and management functionality within `SelectorGroupChat`. This has been raised repeatedly in Discord, Discussions (especially #6347), and even in Help channels with similar requests.

#### **Personal Use Case**  
That said, I’m sharing my perspective here not as a contributor, but as a user who practically needs this functionality.

#### **Limitations of #6350**  
While #6350 does address a similar issue, its TTL cutoff approach simply limits the number of messages. This doesn’t quite meet the need for summarizing or selectively preserving internal messages.

Specifically, in the case of `SelectorGroupChat`, TTL cutoff could potentially remove critical messages, including the initial user request, which raises a concern that the selector might lose context and misidentify the next agent. I am concerned that TTL alone may not address this effectively.

#### **Why model_context Works Better for Me**  
The model_context-based approach proposed in this PR, particularly using `HeadAndTailChatCompletionContext`, allows for reliably preserving both the initial and most recent messages. This ensures that `SelectorGroupChat` can always reference the original user intent when choosing the next speaker, which is essential for the use cases I face. Achieving this kind of context preservation through a simple TTL mechanism seems difficult.

#### **Concern About Expanding #6350 Scope**  
If #6350 were to expand beyond TTL cutoff into more complex message preservation or summarization, it might blur the responsibility between simple message cleanup and full history management. This could make the purpose of each mechanism less clear.

#### **Conclusion**  
Therefore, I personally see #6350 as a clean and focused solution for trimming unnecessary messages, and I’m very supportive of that contribution moving forward. However, this PR enables more precise conversation flow control through internal message summarization and history context management, and it’s something I was also looking forward to seeing merged.

I believe the two are not in conflict—they solve different problems and can complement each other well.

---

#### **Additional Note**  
AutoGen’s model_context structure is already designed to allow users to customize message management without requiring external extensions. That said, tools like the community extension `autogen-contextplus` (which I contributed to) or future model_context improvements could make history management within `SelectorGroupChat` even more flexible and powerful.

ekzhu reviewed (commented):
Let's add some unit tests to show that model context is being managed, validate it using `ReplayChatCompletionClient` which records the calls.

ekzhu commented on python/packages/autogen-agentchat/src/autogen_agentchat/teams/_group_chat/_selector_group_chat.py:
```diff
@@ -173,17 +200,39 @@ async def select_speaker(self, thread: List[BaseAgentEvent | BaseChatMessage]) -
 
-    async def _select_speaker(self, roles: str, participants: List[str], history: str, max_attempts: int) -> str:
+    def construct_message_history(
+        self, message_history: Sequence[Union[BaseChatMessage, BaseAgentEvent, UserMessage, AssistantMessage]]
```
At this point, there shouldn't be any `BaseChatMessage` or `BaseAgentEvent` in the input list right?

ekzhu commented on python/packages/autogen-agentchat/src/autogen_agentchat/teams/_group_chat/_selector_group_chat.py:
```diff
@@ -453,6 +502,7 @@ def __init__(
         candidate_func: Optional[CandidateFuncType] = None,
         custom_message_types: List[type[BaseAgentEvent | BaseChatMessage]] | None = None,
         emit_team_events: bool = False,
+        model_context: ChatCompletionContext | None = None,
```
We need to update the API doc (argument list) and include a code example of using a custom model context.

Ethan0456 replied:
Done! Updated the API doc to include `model_context` in parameter list and added a code example demonstrating its use.

ekzhu commented on python/packages/autogen-agentchat/tests/test_group_chat.py:
```diff
@@ -691,6 +692,50 @@ async def test_selector_group_chat(runtime: AgentRuntime | None) -> None:
+    agent2 = AssistantAgent("agent2", model_client=agent_two_model_client, description="Assistant agent 2")
+
+    termination = TextMentionTermination("TERMINATE")
+    team = SelectorGroupChat(
```
I don't see the model context is being customized here?

Ethan0456 replied:
Done! Updated it to use a prefilled `BufferedChatCompletionContext`.

ekzhu reviewed (changes requested):
Can you resolve the merge conflict with  the main branch?

Ethan0456 commented:
Hi @ekzhu,

I’ve completed the following updates:

- Refactored message history construction to use `LLMMessage` - [2580acf](https://github.com/microsoft/autogen/pull/6330/commits/2580acfccd75b2b3fc82bfabd528320fb82efbad)
- Updated the API documentation to include the `model_context` parameter and added an example - [26f6de0](https://github.com/microsoft/autogen/pull/6330/commits/26f6de0)
- Added a unit test to validate `model_context` usage in `SelectorGroupChat` - [63212ef](https://github.com/microsoft/autogen/pull/6330/commits/63212efddd58ed719c43dd803c96d37ff079127b)
- Integrated `model_context` into `SelectorGroupChatConfig` - [57dbeaa](https://github.com/microsoft/autogen/pull/6330/commits/57dbeaa148223021489ea3cfe9bed3f9aa7310b7)
- Resolved merge conflicts - [6213476](https://github.com/microsoft/autogen/pull/6330/commits/6213476d614debb4477a4ea65d946108a82715c7)

Please let me know if there’s anything else that needs to be addressed.

ekzhu commented on python/packages/autogen-agentchat/src/autogen_agentchat/teams/_group_chat/_selector_group_chat.py:
```diff
@@ -184,17 +206,34 @@ async def select_speaker(self, thread: List[BaseAgentEvent | BaseChatMessage]) -
+
+    async def _select_speaker(self, roles: str, participants: List[str], max_attempts: int) -> str:
+        model_context_messages = await self._model_context.get_messages()
+        model_context_history = self.construct_message_history(model_context_messages)  # type: ignore
```
```suggestion
        model_context_history = self.construct_message_history(model_context_messages)
```

ekzhu commented on python/packages/autogen-agentchat/src/autogen_agentchat/teams/_group_chat/_selector_group_chat.py:
```diff
@@ -184,17 +206,34 @@ async def select_speaker(self, thread: List[BaseAgentEvent | BaseChatMessage]) -
         return agent_name
 
-    async def _select_speaker(self, roles: str, participants: List[str], history: str, max_attempts: int) -> str:
+    def construct_message_history(self, message_history: Sequence[LLMMessage]) -> str:
```
```suggestion
    def construct_message_history(self, message_history: List[LLMMessage]) -> str:
```

Make sure it plays nicely with the ChatCompletionContext

withsmilo commented:
@ekzhu @SongChiYoung @Ethan0456 
It feels like I’ve discovered your passionate discussion a bit too late.

I am a contributor of #6350. While resolving the merge conflict in #6350, I found that this PR has broken my PR. 🤣 

As discussed with @ekzhu in https://github.com/microsoft/autogen/discussions/6169, my goal is to reduce the load which is increased on the external database by cutting message_thread's length. Let me think more about how #6350 can be improved. If you have any good feedback, it is always welcome. Thanks.

appuk45 commented:
Hi Guys,
Can you add any example usages of the `model_context` in `SelectorGroupChat`?. The type of the variable `model_conext` is `ChatCompletionContext`. So I have some doubts regarding the usage. So if you guys can give a small example of that, It will be helpful

Ethan0456 commented:
Hi @appuk45, a small usage example is included in the docstring of the class [here](https://github.com/microsoft/autogen/blob/89927ca43661af0cc1be31deaf7d4da4b5b6fe5e/python/packages/autogen-agentchat/src/autogen_agentchat/teams/_group_chat/_selector_group_chat.py#L515)
